src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts

```diff
@@ -410,36 +423,13 @@ export class ObjectExplorerService implements IObjectExplorerService {
 						allProviders.push(...nodeProviders);
 					}
 
-					self._sessions[session.sessionId!].nodes[node.nodePath].expandEmitter.event((expandResult: NodeExpandInfoWithProviderId) => {
-						if (expandResult && expandResult.providerId) {
-							resultMap.set(expandResult.providerId, expandResult);
-							// If we got an error result back then send error our error event
-							// We only do this for the MSSQL provider
-							if (expandResult.errorMessage && expandResult.providerId === mssqlProviderName) {
-								const errorType = expandResult.errorMessage.indexOf('Object Explorer task didn\'t complete') !== -1 ? 'Timeout' : 'Other';
-								// For folders send the actual name of the folder (since the nodeTypeId isn't useful in this case and the names are controlled by us)
-								const nodeType = node.nodeTypeId === NodeType.Folder ? node.label : node.nodeTypeId;
-								this._telemetryService.createErrorEvent(TelemetryKeys.TelemetryView.Shell, TelemetryKeys.TelemetryError.ObjectExplorerExpandError, undefined, errorType)
-									.withAdditionalProperties({
-										nodeType,
-										providerId: expandResult.providerId
-									}).send();
-							}
-
-						} else {
-							this.logService.error('OE provider returns empty result or providerId');
-						}
+					self._sessions[session.sessionId!].nodes[node.nodePath].expandEmitter.event(
+						(expandResult: NodeExpandInfoWithProviderId) =>
+							resolve(this.completedExpandRequest(expandResult, resultMap, node, newRequest, session, allProviders)));
 
-						// When get all responses from all providers, merge results
-						if (resultMap.size === allProviders.length) {
-							resolve(self.mergeResults(allProviders, resultMap, node.nodePath));
+					self._onExpandCompleteWithoutNodes.event((expandResult: NodeExpandInfoWithProviderId) =>
```

Comment:
This is going to fire regardless of the node path which doesn't seem right. We only want to be watching for expand completed events for this node specifically...what happens if another node of the same provider is also expanding and triggers this event while this one is still waiting for the responses?

Reply:
You mean this session - not to confuse 'node' with session.nodes. We have determined there are no nodes available, when we call this. Can you even expand the same session again while waiting for it?

Reply:
So the _onExpandCompleteWithoutNodes event is global to the service. It currently is fired for EVERY expandResponse that doesn't have a session found.

Then here you're setting up the handlers for specific nodes - which will complete when it gets the expand responses from all providers for that node. 

But because you're hooking into the global _onExpandCompleteWithoutNodes event then every node that's currently expanding will have this triggered when that event is triggered. Which then calls completedExpandRequest. But that function doesn't know its for a specific node - for example it immediately sets the resultMap for the provider of the expandResult. But that expandResult may have come from some other random session that fired the _onExpandCompleteWithoutNodes event. 

At the very least you should only be handling that event trigger if it's for the session the node was originally for. 

(but of course all this stuff is just hiding what seems to be a more serious issue to me - why can it not find the session in the first place in the onNodeExpanded function)?